modules/apps/29-fee/keeper/keeper.go

```diff
@@ -0,0 +1,307 @@
+	return k.portKeeper.BindPort(ctx, portID)
+}
+
+// GetChannel wraps IBC ChannelKeeper's GetChannel function
+func (k Keeper) GetChannel(ctx sdk.Context, portID, channelID string) (channeltypes.Channel, bool) {
+	return k.channelKeeper.GetChannel(ctx, portID, channelID)
+}
+
+// GetNextSequenceSend wraps IBC ChannelKeeper's GetNextSequenceSend function
+func (k Keeper) GetNextSequenceSend(ctx sdk.Context, portID, channelID string) (uint64, bool) {
+	return k.channelKeeper.GetNextSequenceSend(ctx, portID, channelID)
+}
+
+// GetFeeAccount returns the ICS29 Fee ModuleAccount address
+func (k Keeper) GetFeeModuleAddress() sdk.AccAddress {
+	return k.authKeeper.GetModuleAddress(types.ModuleName)
+}
+
+// SetFeeEnabled sets a flag to determine if fee handling logic should run for the given channel
+// identified by channel and port identifiers.
+func (k Keeper) SetFeeEnabled(ctx sdk.Context, portID, channelID string) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.FeeEnabledKey(portID, channelID), []byte{1})
+}
+
+// DeleteFeeEnabled deletes the fee enabled flag for a given portID and channelID
+func (k Keeper) DeleteFeeEnabled(ctx sdk.Context, portID, channelID string) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.FeeEnabledKey(portID, channelID))
+}
+
+// IsFeeEnabled returns whether fee handling logic should be run for the given port. It will check the
+// fee enabled flag for the given port and channel identifiers
+func (k Keeper) IsFeeEnabled(ctx sdk.Context, portID, channelID string) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Get(types.FeeEnabledKey(portID, channelID)) != nil
+}
+
+// GetAllFeeEnabledChannels returns a list of all ics29 enabled channels containing portID & channelID that are stored in state
+func (k Keeper) GetAllFeeEnabledChannels(ctx sdk.Context) []*types.FeeEnabledChannel {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(types.FeeEnabledKeyPrefix))
+	defer iterator.Close()
+
+	var enabledChArr []*types.FeeEnabledChannel
+	for ; iterator.Valid(); iterator.Next() {
+		keySplit := strings.Split(string(iterator.Key()), "/")
+
+		ch := &types.FeeEnabledChannel{
+			PortId:    keySplit[1],
+			ChannelId: keySplit[2],
+		}
+
+		enabledChArr = append(enabledChArr, ch)
+	}
+
+	return enabledChArr
+}
+
+// DisableAllChannels will disable the fee module for all channels.
+// Only called if the module enters into an invalid state
+// e.g. ModuleAccount has insufficient balance to refund users.
+// In this case, chain developers should investigate the issue, fix it,
+// and then re-enable the fee module in a coordinated upgrade.
+func (k Keeper) DisableAllChannels(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(types.FeeEnabledKeyPrefix))
+
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		store.Delete(iterator.Key())
```

Comment:
Great catch! This is a bug. It is not safe to delete key while iterating over it. #780 

As a side note, there should be a linter for this. cc @crodriguezvega this could one of the patterns for the process trail of bits proposed (I forget what it was called)

Reply:
Yes, [semgrep](https://semgrep.dev). We could maybe use it for this indeed.